File: channel/receive.cc
#include "channel.ih"

template<typename T>
T Channel<T>::receive() {
    unique_lock<mutex> ul(d_mutex);
    ++d_receivers_present;

    d_senders.notify_all();         // Notify senders I'm here
    while (size() == 0) {           // Wait until item
        d_receivers.wait(ul);
    }
    // Locked; there is an item

    T result(move(front()));        // Take the item
    pop_front();
    --d_receivers_present;
    return result;
}

File: channel/receive.f
template<typename T>
void Channel<T>::receive(T *result) {
    std::unique_lock<std::mutex> ul(d_mutex);
    ++d_receivers_present;

    try {
        waitForSenders(ul);
        // Locked; there is an item
        moveElement(result);
        --d_receivers_present;
    } catch (...) {
        --d_receivers_present;
        throw;
    }
}